packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -994,9 +1098,23 @@ function beginWork(
       case HostComponent:
         pushHostContext(workInProgress);
         break;
-      case ClassComponent:
-        pushLegacyContextProvider(workInProgress);
+      case ClassComponent: {
+        const Component = workInProgress.type;
+        const childContextTypes = Component.childContextTypes;
+        if (childContextTypes !== null && childContextTypes !== undefined) {
+          pushLegacyContextProvider(workInProgress);
+        }
+        break;
+      }
+      case ClassComponentLazy: {
+        const thenable = workInProgress.type;
+        const Component = (thenable._reactResult: any);
```

Comment:
Can we wrap all reads of `_reactResult` in a helper. Like `readResolvedType(workInProgress)`. Normally I wouldn't recommend that but in this case it is a user space object. We might want to add assertions or warnings in the future. It is also a indicator that we can use that this is external to Fiber and will have inconsistent hidden classes. It might just be a a few hidden classes and it's good that those optimizations get to use the same inline cache if possible rather than redoing that work at every callsite.

Reply:
What about ReactTestRenderer?

Reply:
Test renderer doesn't need to.